### DISCUSSION TITLE
feat(attributes): Allow for custom attributes

### DISCUSSION THREAD

**Discussion by Maddiaa0** [noir-lang/noir#3412] [Ideas]
### Problem

In my perfect world, noir would support for macros, a way to get there is to allow for arbitrary attibutes names that are not restricted to some built in functionality like blackbox or oracle calls. 

e.g.
```rust
#[some_macro]
fn this_function_is_transformed_into_a_macro() {
    ...
}
```

Here is a hacky alternative I tried to create on a branch https://github.com/noir-lang/noir/commit/2550c953d60867abae7abe3239e4e68942141a23. Which has a purpose of creating a pseudo macro system (warning its complete spaghetti code)

Here is the diff in aztec library : https://github.com/AztecProtocol/aztec-packages/blob/b007330e24bb18932acd75f39b25efed1d1959aa/yarn-project/noir-contracts/src/contracts/private_token_contract/src/main.nr

### Alternatives Considered

-

### Additional Context

-

### Would you like to submit a PR for this Issue?

No

### Support Needs

-

**Comment by jfecher** [comment]
This issue is requesting attributes but seems to more deeply be asking for macros in some form. These are not necessarily related so it may be helpful to narrow things down some more:
- What functionality would you want for macros? 
  - E.g Purely just `macro_rules!` style code expansion and/or procedural macros? If procedural macros, do you really want procedural macros or just better forms of meta programming in general?
  - What are some use cases you need meta programming for?
  - What meta programming features would be required by these usecases?
- What functionality would you want for attributes?

Meta programming can be a topic that has sweeping implications for tooling and code readability so I think it is helpful to identify some concrete use cases and boundaries for it.

I also think meta programming (like lifetimes) is one place where we should diverge from rust. Proc-macros in general are far too complex and if we need something similar I'd prefer to swap them out wholesale with another system without the same limitations/complexity.
